internal/logsink/logsink_test.go

```diff
@@ -0,0 +1,673 @@
+		}(i, entries[i*amount:(i*amount)+amount])
+	}
+
+	// Wait for all the flushes to complete.
+	s.expectNumOfFlushes(c, division)
+
+	lines := parseLog(c, buffer)
+	c.Assert(lines, gc.HasLen, total, gc.Commentf("expected %d lines, got %d", total, len(lines)))
+
+	expected := make([]logRecord, total)
+	for k, entry := range entries {
+		expected[k] = logRecord{
+			Time:     entry.Time,
+			Level:    entry.Level.String(),
+			Message:  entry.Message,
+			Module:   entry.Module,
+			Location: entry.Location,
+			Labels: map[string]string{
+				"foo": "bar",
+			},
+			ModelUUID: "uuid",
+		}
+	}
+
+	// We can't guarantee the order of the entries written in the test, so we
+	// need to sort them before comparing.
+	sort.Slice(lines, func(i, j int) bool {
+		return lines[i].Time.Before(lines[j].Time)
+	})
+
+	c.Check(lines, gc.DeepEquals, expected)
+
+	workertest.CleanKill(c, sink)
+}
+
+func (s *logSinkSuite) TestLogAndWriteInterleaved(c *gc.C) {
+	// Send more than two batches of logs, but less than the batch size.
+	// This will force two flushes and an additional tick and a flush.
+
+	sink, buffer := s.newLogSink(c, 50)
+	defer workertest.DirtyKill(c, sink)
+
+	total := (rand.Intn(48) + 1) + 100
+
+	now := time.Now().UTC()
+
+	entries := make([]loggo.Entry, total)
+	for i := range total {
+		entries[i] = loggo.Entry{
+			Timestamp: now.Add(time.Duration(i) * time.Second),
+			Level:     loggo.INFO,
+			Message:   fmt.Sprintf("hello-%d", i),
+			Module:    "module",
+			Filename:  "file.go",
+			Line:      i,
+			Labels: map[string]string{
+				"model-uuid": "uuid",
+			},
+		}
+	}
+
+	for i, entry := range entries {
+		if i%2 == 0 {
+			sink.Write(entry)
+		} else {
+			sink.Log([]logger.LogRecord{{
+				Time:      entry.Timestamp,
+				ModelUUID: "uuid",
+				Module:    entry.Module,
+				Location:  fmt.Sprintf("%s:%d", entry.Filename, entry.Line),
+				Level:     logger.Level(entry.Level),
+				Message:   entry.Message,
+				Labels:    entry.Labels,
+			}})
+		}
+	}
+
+	// We should see 2 flushes, and flush via the remaining entries.
+	s.expectNumOfFlushes(c, 2)
+	s.expectTick(c)
+	s.expectFlush(c)
+
+	lines := parseLog(c, buffer)
+	c.Assert(lines, gc.HasLen, total, gc.Commentf("expected %d lines, got %d", total, len(lines)))
+
+	expected := make([]logRecord, total)
+	for k, entry := range entries {
+		expected[k] = logRecord{
+			Time:     entry.Timestamp,
+			Level:    entry.Level.String(),
+			Message:  entry.Message,
+			Module:   entry.Module,
+			Location: fmt.Sprintf("%s:%d", entry.Filename, entry.Line),
+			Labels: map[string]string{
+				"model-uuid": "uuid",
+			},
+			ModelUUID: "uuid",
+		}
+	}
+	c.Check(lines, gc.DeepEquals, expected)
+
+	workertest.CleanKill(c, sink)
+}
+
+func (s *logSinkSuite) newLogSink(c *gc.C, batchSize int) (*LogSink, *bytes.Buffer) {
+	s.states = make(chan string, 1)
+
+	buffer := new(bytes.Buffer)
+
+	sink := newLogSink(buffer, batchSize, time.Millisecond*100, s.states)
+	return sink, buffer
+}
```

Comment:
We need at least 1 test that does not pass the states channel. The channel will add some level of synchronisation and may cause the race tests to pass when a race exists in the code.

I don't see a race here, but implementations can change and testing feedback channels can cause differences.